### PR TITLE
[spirv] SpirvBuilder function and basic block methods.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

**Comment by antiagainst** [comment]
Wow, this is quick! ;)

**Comment by ehsannas** [comment]
This is only the Function and BasicBlock methods... Still have to flesh out instruction insertion methods.

File: tools/clang/lib/SPIRV/SpirvBuilder.cpp [hunk 1]
```diff
@@ -17,5 +17,100 @@ SpirvBuilder::SpirvBuilder(SpirvContext &ctx)
...
+                                           SourceLocation loc,
+                                           llvm::StringRef funcName) {
+  assert(!function && "found nested function");
+
+  // Create the function
+  function = new (context) SpirvFunction(
+      returnType, /*id*/ 0, spv::FunctionControlMask::MaskNone, loc);
+
+  // Add a debug name for the function
+  auto *debugName = new (context) SpirvName(loc, function, funcName);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SpirvBuilder.cpp]: Hmm, I feel we should have a `name` field baked into `SpirvInstruction` directly and provide API to set there.

File: tools/clang/include/clang/SPIRV/EmitVisitor.h [hunk 2]
```diff
@@ -90,10 +89,31 @@ class EmitVisitor : public Visitor {
...
+  std::vector<uint32_t> preambleBinary;
+  // All debug instructions *except* OpLine. Includes:
+  // OpString, OpSourceExtension, OpSource, OpSourceContinued, OpName,
+  // OpMemberName, OpModuleProcessed
+  std::vector<uint32_t> debugBinary;
+  // All annotation instructions: OpDecorate, OpMemberDecorate, OpGroupDecorate,
+  // OpGroupMemberDecorate, and OpDecorationGroup.
+  std::vector<uint32_t> annotationsBinary;
+  // All OpLine instructions
+  std::vector<uint32_t> lineInfoBinary;
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/EmitVisitor.h]: This should be the `typeConstantBinary`?

File: tools/clang/include/clang/SPIRV/SpirvFunction.h [hunk 3]
```diff
@@ -50,6 +50,9 @@ class SpirvFunction {
   uint32_t getReturnTypeId() const { return 0; }
   uint32_t getFunctionTypeId() const { return 0; }
 
+  void setFunctionName(llvm::StringRef name) { functionName = name; }
+  llvm::StringRef getFunctionName() { return functionName; }
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/SpirvFunction.h]: const

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h [hunk 4]
```diff
@@ -120,6 +120,9 @@ class SpirvInstruction {
 
   clang::SourceLocation getSourceLocation() const { return srcLoc; }
 
+  void setDebugName(llvm::StringRef name) { debugName = name; }
+  llvm::StringRef getDebugName() { return debugName; }
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/SpirvInstruction.h]: const

File: tools/clang/lib/SPIRV/EmitVisitor.cpp [hunk 5]
```diff
@@ -17,14 +17,59 @@
...
-  spirvBinary.insert(spirvBinary.end(), curInst.begin(), curInst.end());
+  switch (op) {
+  case spv::Op::OpCapability:
+  case spv::Op::OpExtension:
+  case spv::Op::OpExtInstImport:
+  case spv::Op::OpMemoryModel:
+  case spv::Op::OpEntryPoint:
+  case spv::Op::OpExecutionMode:
+  case spv::Op::OpExecutionModeId:
+    preambleBinary.insert(preambleBinary.end(), curInst.begin(), curInst.end());
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitVisitor.cpp]: Need break

File: tools/clang/lib/SPIRV/EmitVisitor.cpp [hunk 6]
```diff
@@ -17,14 +17,59 @@
...
+  case spv::Op::OpDecorate:
+  case spv::Op::OpDecorateId:
+  case spv::Op::OpMemberDecorate:
+  case spv::Op::OpGroupDecorate:
+  case spv::Op::OpGroupMemberDecorate:
+  case spv::Op::OpDecorationGroup:
+  case spv::Op::OpDecorateStringGOOGLE:
+  case spv::Op::OpMemberDecorateStringGOOGLE:
+    annotationsBinary.insert(annotationsBinary.end(), curInst.begin(), curInst.end());
+  case spv::Op::OpLine:
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitVisitor.cpp]: OpLine can be everywhere. They should just go to the `default` case.

File: tools/clang/include/clang/SPIRV/EmitVisitor.h (hunk 2 above)
Comment (ehsannas) [reply tools/clang/include/clang/SPIRV/EmitVisitor.h]: Done

File: tools/clang/include/clang/SPIRV/SpirvFunction.h (hunk 3 above)
Comment (ehsannas) [reply tools/clang/include/clang/SPIRV/SpirvFunction.h]: Done

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h (hunk 4 above)
Comment (ehsannas) [reply tools/clang/include/clang/SPIRV/SpirvInstruction.h]: Done

File: tools/clang/lib/SPIRV/EmitVisitor.cpp (hunk 5 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/EmitVisitor.cpp]: Oooops. Fixed.

File: tools/clang/lib/SPIRV/EmitVisitor.cpp (hunk 6 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/EmitVisitor.cpp]: Done

**Pull request merged**